redbpf-macros/src/lib.rs

```diff
@@ -435,3 +437,122 @@ pub fn tc_action(attrs: TokenStream, item: TokenStream) -> TokenStream {
 
     probe_impl("tc_action", attrs, wrapper, name)
 }
+
+/// Define a tracepoint.
+#[proc_macro_attribute]
+pub fn tracepoint(attrs: TokenStream, item: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(item as ItemFn);
+    let ident = item.sig.ident.clone();
+
+    let (name, split) = if !attrs.is_empty() {
+        match parse_macro_input!(attrs as Expr) {
+            Expr::Lit(ExprLit {
+                lit: Lit::Str(s), ..
+            }) => (s.value(), ":"),
+            _ => panic!("expected string literal"),
+        }
+    } else {
+        (ident.to_string(), "__")
+    };
+    let mut iter = name.split(split);
+    let category = iter.next().expect("category");
+    let name = iter.next().expect("name");
+    assert_eq!(iter.next(), None);
+
+    let section_name = format!("tracepoint/{}__{}", category, name);
+    let outer_name = format_ident!("tracepoint__{}__{}", category, name);
+    let struct_ident = format_ident!("{}", name.to_camel_case());
+    let event_struct = gen_event_struct(category, name, &struct_ident);
+
+    let tokens = quote! {
+        #event_struct
+
+        #[no_mangle]
+        #[link_section = #section_name]
+        fn #outer_name(args: *const core::ffi::c_void) -> i32 {
+            #item
+
+            #ident(unsafe { &*(args as *const #struct_ident) });
+            return 0;
+        }
+    };
+
+    tokens.into()
+}
+
+fn gen_event_struct(category: &str, name: &str, ident: &syn::Ident) -> TokenStream2 {
```

Comment:
This function could use some more vertical white space, and a few comments about what is exactly happening. It would also be helpful to add some examples of the lines that are being parsed just to get an idea about the data format.